src/components/colorscale/calc.js

```diff
@@ -20,7 +20,7 @@ module.exports = function calc(trace, vals, containerStr, cLetter) {
 
     if(containerStr) {
         container = Lib.nestedProperty(trace, containerStr).get();
-        inputContainer = Lib.nestedProperty(trace._input, containerStr).get();
+        inputContainer = Lib.nestedProperty(trace, containerStr).get();
```

Comment:
@etpinard 

> In what scenario? `calcColorscale` is only called if `hasMarkers(trace)` is true (and if `hasLines(trace)`, no ?

From a previous comment: 

> The technical difficulty is that this object, on which it attempts to set `cmin`/`cmax` may not exist: marker inherits props to marker.line but marker.line might not have been specified among the input. Therefore the code attempts to set `cmin`/`cmax` on `undefined`.

> I really can't see a reason for calcCalcscale to be modified in this PR. Could you elaborate?

In the past, the color determination was much simpler for `scatter3d`: its lines, markers did *not* participate in the sophisticated color specification that has the attributes from `color_attributes.js`. Therefore this specific piece of code was not exercised for `scatter3d` lines, markers and marker lines. This explains why there was no issue with `cmin`/`cmax` here (didn't get called and these props weren't present for `scatter3d`).

It looked sensible to use common code (`colorscale/calc.js`) to deal with the common color attribute set (`color_attributes.js`). Therefore `colorscale/calc.js` got exercised in new ways, under which it failed (in my understanding) for the following two reasons (not new, just a recap of previous discussion):

1. It [errors out](https://github.com/plotly/plotly.js/pull/617/files/dc70494fee7dd21bc6255dcddf34381fec549bee#r67069588) because it tries to set attributes on an object that may not exist (e.g. `marker.line`)
2. Persisting a *calculated* minimum/maximum value means that if the user sets `cauto` to `true`, then back to `false`, then [they will get a different result](https://github.com/plotly/plotly.js/pull/617/files/dc70494fee7dd21bc6255dcddf34381fec549bee#r67126182) than what they originally had (`cmin`/`cmax` would get overwritten by the data-drivem minimum/maximum).

So I thought we should use this piece of code for the newly introduced `scatter3d` coloring options for `line` and `marker` rather than duplicating this file and it implies a change. Alternatively, if you'd like, I can copy this file just for `scatter3d` and contidionally invoke that. Another alternative is if I use the [above snippet](https://github.com/plotly/plotly.js/pull/617/files/dc70494fee7dd21bc6255dcddf34381fec549bee#r67069588) to solve the case of missing object, so that the tests pass, but in this case, as mentioned, toggling `cauto` back and forth will override the original user-supplied `cmin`/`cmax` values.